walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,180 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
+  private int meetingDuration;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // The total (common) free time we have to conduct the meeting.
+    LinkedList<TimeRange> freeTime = new LinkedList<TimeRange>();
+    freeTime.addAll(Arrays.asList(TimeRange.WHOLE_DAY));
+
+    meetingDuration = (int) request.getDuration();
+    LinkedList<String> optionalAttendees = new LinkedList<>(request.getOptionalAttendees());
+
+    // If duration of the meeting is greater than a whole day, then return an empty list.
+    if (meetingDuration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If there are no attendees then return the whole day (ie the free time we have till now).
+    if (request.getAttendees().isEmpty() && optionalAttendees.isEmpty()) {
+      return freeTime;
+    }
+
+    // To store all the events which the optional attendees are attending.
+    LinkedList<Event> optionalAttendeeEvents = new LinkedList<>();
+
+    /** Changes freeTime such that it now conatins all the free time slots when the meeting can
+     * happen, updating the events that optional attendees can attend.
+     */
+    freeSlotsWithMandatoryAttendees(
+        freeTime, events, optionalAttendeeEvents, optionalAttendees, request);
+
+    /** If free time is not empty and optional attendees are not empty
+     * then go ahead and look for time slots to have meeting with our optional attendees 
+     */
+    if (!freeTime.isEmpty() && !optionalAttendeeEvents.isEmpty()) {
+      freeTime =
+          freeSlotsWithOptionalAttendees(optionalAttendeeEvents, freeTime, optionalAttendees);
+    }
+
+    return freeTime;
+  }
+
+  // Loop through all the events given and modify freeTime List
+  private void freeSlotsWithMandatoryAttendees(
+      LinkedList<TimeRange> freeTime,
+      Collection<Event> events,
+      LinkedList<Event> optionalAttendeeEvents,
+      LinkedList<String> optionalAttendees,
+      MeetingRequest request) {
+
+    for (Event event : events) {
+      /* Check if the event has any optional attendees and
+       * save it in the optionalAtttendeeEvents list which we will deal with
+       * later in the freeSlotsWithOptionalAttendees method
+       */
+      if (!Collections.disjoint(optionalAttendees, event.getAttendees()))
```

Comment:
Maybe I don't understand the logic here, but I think you want to call `findConflictAndResolve` if `request.getAttendees()` and `event.getAttendees()` have a non-empty intersection. *Otherwise*, you want to add to the `optionalAttendeeEvents`, if `optionalAttendees` and `event.getAttendees()` are not empty.

Reply:
If you use this logic, you can remove the additional attendees check in `findConflictAndResolve`.

Reply:
That is right, I am aiming to do that.
And I kept the check in findConflictAndResolve especially for the time when we call this function for mandatory attendees, to see if there are any mandatory attendees attending that event
(as we loop through different events, there can be events in which neither our mandatory nor optional attendees are present).
Hence, it also uses it for optionalAttendeeEvents again.

Reply:
Um if I think about it again, if I add an 
if (! Collections.disjoint(mandatoryAttendees, event.getAttendees()))
instead of 'else' below, then I think it'll do the job and I don't need the check in findConflictAndResolve neither do I need to pass the attendees list to it.
Thanks for pointing it out :)
Let me know if this is right.

Reply:
The code you have works fine, but resolves some meeting conflicts twice: If an event contains 1 mandatory and 1 optional attendee, you should call `findConflictAndResolve` and you don't have to add it to`optionalAttendeeEvents`.

Reply:
Yes, noted that.
I'll add it as an else if then.